File: earnings-call-analyzer/eca/ingest.py
"""Ingest: normalize a transcript source into clean, line-numbered markdown.

Stable line numbers are the backbone of the grounding layer: every claim in the
report cites the line(s) it came from, so the numbering must be deterministic.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class Transcript:
    """A normalized, line-numbered transcript."""

    source: str
    lines: list[str]                       # 1-indexed content: lines[0] is [L1]
    exhibits: list[str] = field(default_factory=list)  # flagged images/exhibits

    def numbered_markdown(self) -> str:
        """Render as `[Ln] text` — the exact form the model reads and cites."""
        return "\n".join(f"[L{i}] {t}" for i, t in enumerate(self.lines, start=1))

    def line(self, n: int) -> str | None:
        """Return the text of line n (1-indexed), or None if out of range."""
        if 1 <= n <= len(self.lines):
            return self.lines[n - 1]
        return None


_IMAGE_RE = re.compile(r"!\[[^\]]*\]\([^)]*\)")
_EXHIBIT_RE = re.compile(r"\b(exhibit|figure|chart|table|appendix)\s+[A-Z0-9]", re.IGNORECASE)


def ingest(path: str | Path) -> Transcript:
    """Load a .txt/.md/.pdf/.docx/.pptx transcript and normalize it into a Transcript."""
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"Transcript not found: {path}")

    from .convert import convert_to_markdown
    raw = convert_to_markdown(path)
    lines: list[str] = []
    exhibits: list[str] = []

    for raw_line in raw.splitlines():
        text = raw_line.strip()
        if not text:
            continue  # collapse blank lines so numbering tracks meaningful content
        if _IMAGE_RE.search(text) or _EXHIBIT_RE.search(text):
            exhibits.append(text)
        # keep the line in the body regardless; exhibits are flagged, not removed
        # normalize internal whitespace runs to single spaces for stable numbering
        text = re.sub(r"[ \t]+", " ", text)
        lines.append(text)

    return Transcript(source=str(path), lines=lines, exhibits=exhibits)

File: earnings-call-analyzer/web/src/ingest.ts
// Ingest: normalize raw transcript text into stable, line-numbered content.
// Stable line numbers are the backbone of grounding: every claim cites its [Ln].
// (Text-only in the browser/Worker; PDF extraction stays in the Python reference tool.)

export interface Transcript {
  lines: string[]; // 1-indexed content: lines[0] is [L1]
  exhibits: string[];
}

const IMAGE_RE = /!\[[^\]]*\]\([^)]*\)/;
const EXHIBIT_RE = /\b(exhibit|figure|chart|table|appendix)\s+[A-Z0-9]/i;

export function ingest(raw: string): Transcript {
  const lines: string[] = [];
  const exhibits: string[] = [];
  for (const rawLine of raw.split(/\r?\n/)) {
    let text = rawLine.trim();
    if (!text) continue; // collapse blanks so numbering tracks meaningful content
    if (IMAGE_RE.test(text) || EXHIBIT_RE.test(text)) exhibits.push(text);
    text = text.replace(/[ \t]+/g, " ");
    lines.push(text);
  }
  return { lines, exhibits };
}

export function numberedMarkdown(t: Transcript): string {
  return t.lines.map((text, i) => `[L${i + 1}] ${text}`).join("\n");
}

export function line(t: Transcript, n: number): string | null {
  return n >= 1 && n <= t.lines.length ? t.lines[n - 1] : null;
}
